### PR TITLE
Installdb.php additional options

### PR DESCRIPTION
**Pull request by xlecours**
_This is an almost exact copy of Justin's https://github.com/aces/Loris/pull/2723
Thanks for this contribution!_

`installdb.php` now allows you to:

+ Specify a database and mark it as already created (so the script doesn't try to create something that already exists)
+ Specify that the database was already created
+ Specify that the config is already good
+ Specify that the MySQL user was already created

Also fixed a small part where it tried to access `mysql.users` to check if you can `CREATE DATABASE`. Not necessary. It's entirely possible that you can `CREATE DATABASE` but cannot `SELECT ... FROM mysql.users`.

This is needed for us, especially, and for the summer interns coming in. They'll be installing Loris with this script and they will not be able to `CREATE DATABASE` or `CREATE USER` because an empty database and users will already be provided by `[insert whoever is responsible for that]`.

### CODE REVIEW

File: smarty/templates/install.tpl [hunk 1]
```diff
@@ -43,19 +43,30 @@
...
+				<label for="lorismysql_already_created">Already Created:
+                <input type="checkbox" name="lorismysql_already_created" value="on" {($lorismysql_already_created == 'on') ? 'checked' : ''}/>
 			</div>
 			</fieldset>
+            <hr/>
+            <div>
+                Check "<strong>Already Created</strong>" only if the appropriate mysql user has already been created!<br/>
+                If you or the web admin wish to give the appropriate privileges to an existing MySQL user, <br/>
+                the privileges are <code>SELECT, INSERT, UPDATE, DELETE, CREATE, DROP, REFERENCES, CREATE TEMPORARY TABLES, LOCK TABLES</code><br/>
+                for the following: <code>`{$dbname}`.*</code>
```
Comment (driusan) [inline smarty/templates/install.tpl]: I don't think the "If you or the web admin" .. sentence belongs here. It's unlikely to be properly maintained in a template in the installer and anyone who's bypassing the web based installer should already know enough about what they're doing to figure that out from elsewhere.

File: smarty/templates/install.tpl [hunk 2]
```diff
@@ -151,6 +165,41 @@
...
 			</div>
+            <hr/>
+			<div>
+				<div class="col-md-2">
+					<label for="use_existing_database">Use Existing Database:</label>
+				</div>
+				<div class="col-md-9">
+					<input type="checkbox" id="use_existing_database" name="use_existing_database" value="on" {($use_existing_database == 'on') ? 'checked' : ''}/>
+                    Check this if the database already exists but does not have Loris installed yet
+                    <!--The following is a hack to get the UI to align correctly *shrug*-->
```
Comment (driusan) [inline smarty/templates/install.tpl]: This should be taken out. (The hack if possible, and the "*shrug*" if fixing the hack is more effort than it's worth.)

(Same goes for the other two instances below.)

File: htdocs/installdb.php [hunk 3]
```diff
@@ -74,7 +74,7 @@
     exit(3);
 }
 
-require_once __DIR__ . "/../vendor/autoload.php";
+require_once __DIR__ . '/../vendor/autoload.php';
```
[the remark this answers is not in the document]
Comment (xlecours) [reply htdocs/installdb.php]: It is for uniformity in the code. Considering the string a literal, there is no justification for using double quotes.

[the remark this answers is not in the document]
Comment (driusan) [reply htdocs/installdb.php]: According to `git grep autoload.php`, the code nearly uniformly uses double quotes, not single quotes for autoload.php. The only exceptions are the api code and the dataquery module code (and I'm not even sure why the dataquery ajax scripts are requiring it at all, since it should have already been by AjaxHelper.php before it gets to that point..)

[the remark this answers is not in the document]
Comment (xlecours) [reply htdocs/installdb.php]:
```
     38 ';
     47 ";
      1 ";
      1 ");
```

[the remark this answers is not in the document]
Comment (xlecours) [reply htdocs/installdb.php]: Nonetheless, I was referring to that file quotation uniformity

File: php/installer/Database.class.inc [hunk 4]
```diff
@@ -44,6 +44,83 @@ class Database
...
+    {
+        $stmt = $this->PDO->prepare("SHOW DATABASES LIKE :v_db");
+        $stmt->execute(array('v_db' => $database ));
+        $exists = !empty($stmt->fetchAll());
+
+        // this needs to be !empty, because otherwise if there's
+        // no results returned count($dbexists) == 1 when no
+        // results are returned, because the result isn't an array
+        // Something to do with how the Countable interface is
+        // implemented in PHP..
```
Comment (driusan) [inline php/installer/Database.class.inc]: This comment is on the wrong line since you moved the empty check above it.

(Unrelatedly, I suspect the "Something to do with how the Countable interface is implemented" comment is referring to is this behaviour:  https://wiki.php.net/rfc/counting_non_countables which was changed in 7.2.

"There are no known use cases for this, and it can easily hide bugs")

File: smarty/templates/install.tpl [hunk 5]
```diff
@@ -3,8 +3,8 @@
 	{* installdb.php is in the root directory, so we know these relative
 	   links should work. Load in the base LORIS bootstrap CSS
 	*}
-		<link rel="stylesheet" href="bootstrap/css/bootstrap.min.css">
-		<link rel="stylesheet" href="bootstrap/css/custom-css.css">
+		<link rel="stylesheet" href="/bootstrap/css/bootstrap.min.css">
+		<link rel="stylesheet" href="/bootstrap/css/custom-css.css">
```
Comment (driusan) [inline smarty/templates/install.tpl]: What's the reason for this change? It's going to make the installer break if LORIS isn't hosted on the root of the web server. (It also makes the comment above it confusing..)

Comment (xlecours) [reply smarty/templates/install.tpl]: The page wasn't loading css otherwise. I will look at using baseurl if possible

File: php/installer/Database.class.inc (hunk 4 above)
Comment (xlecours) [reply php/installer/Database.class.inc]: The comment is just confusing. Removed

File: htdocs/installdb.php [hunk 6]
```diff
@@ -95,31 +95,61 @@
     // This will connect to MySQL, check the permissions of the
     // account provided, check that the database doesn't already
     // exist, and create the database.
-    if ($installer->CreateMySQLDB($_POST) === false) {
-        $tpl_data['error'] = $installer->GetLastError();
-        $tpl_data['Page']  = "";
-        break;
+    if (!isset($_POST['do_not_install'])) {
```
[the remark this answers is not in the document]
Comment (xlecours) [reply htdocs/installdb.php]: `do_not_not_install` ?  I will work on that

File: php/installer/Database.class.inc (hunk 4 above)
Comment (driusan) [reply php/installer/Database.class.inc]: I think the behaviour of count on objects in PHP was super confusing and the comment was trying to explain the workaround it was using..

**Comment by xlecours** [comment]
Ok, I have tested it a couple of times with Raisinbread already sourced and it work just fine now.

**Comment by driusan** [comment]
@ridz1208 @xlecours Code looks fine, but should this go to minor or major?

**Comment by ridz1208** [comment]
@driusan given the alterations to the DB class
i would say major
unless these changes do not affect any other code ?

**Comment by driusan** [comment]
It's `php/installer/Database.class.inc`, not `php/libraries/Database.class.inc`, but even without that it seems like a pretty big change.

**Comment by ridz1208** [comment]
@driusan if its big but doesnt affect anything other than installer
than someone should thoroughly test the installer and thats it
I dnt see the need for a loris testing round where the installer might even be less tested due to the other 155 things to do...
